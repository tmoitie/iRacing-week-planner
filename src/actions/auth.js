import {
  getAuth, signOut as fbSignOut, signInWithEmailAndPassword, createUserWithEmailAndPassword,
  onAuthStateChanged, sendPasswordResetEmail, GoogleAuthProvider, signInWithPopup,
} from 'firebase/auth';
import { debouncedDispatcherSaveSettings, getSettingsFromFirebase, saveSettingsToFirebase } from './settings';

export const LOADING_AUTH = 'AUTH/LOADING_SIGN_IN';
export const ERROR_AUTH = 'AUTH/ERROR_AUTH';
export const ERROR_RESET = 'AUTH/ERROR_RESET';
export const ERROR_ACKNOWLEDGE = 'AUTH/ERROR_ACKNOWLEDGE';
export const SIGNED_IN = 'AUTH/SIGNED_IN';
export const SIGNED_OUT = 'AUTH/SIGNED_OUT';
export const RESET_SENT = 'AUTH/RESET_SENT';
export const LOADING_RESET = 'AUTH/LOADING_RESET';

export function signOut() {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    if (getState().settings.firebaseSynced === false) {
      debouncedDispatcherSaveSettings.cancel();
      await dispatch(saveSettingsToFirebase());
    }
    const auth = getAuth(getState().auth.firebaseApp);
    await fbSignOut(auth);

    dispatch({ type: SIGNED_OUT });
  };
}

export function signIn(email, password) {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    try {
      const auth = getAuth(getState().auth.firebaseApp);
      await signInWithEmailAndPassword(auth, email, password);
      return {};
    } catch (error) {
      dispatch({ type: ERROR_AUTH, error });
      return { type: ERROR_AUTH };
    }
  };
}

export function signInWithGoogle() {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    try {
      const auth = getAuth(getState().auth.firebaseApp);

      const provider = new GoogleAuthProvider();
      provider.addScope('profile');
      provider.addScope('email');

      await signInWithPopup(auth, provider);
      return {};
    } catch (error) {
      dispatch({ type: ERROR_AUTH, error });
      return { type: ERROR_AUTH };
    }
  };
}

export function createAccount(email, password) {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    try {
      const auth = getAuth(getState().auth.firebaseApp);
      await createUserWithEmailAndPassword(auth, email, password);
      return {};
    } catch (error) {
      dispatch({ type: ERROR_AUTH, error });
      return { type: ERROR_AUTH };
    }
  };
}

export function acknowledgeAuthError() {
  return { type: ERROR_ACKNOWLEDGE };
}

export function signedIn(user) {
  return async (dispatch) => {
    await dispatch({ type: SIGNED_IN, user });

    if (user) {
      await dispatch(getSettingsFromFirebase());
    }
  };
}

export function startListener() {
  return async (dispatch, getState) => {
    const auth = getAuth(getState().auth.firebaseApp);
    const { currentUser } = auth;
    dispatch(signedIn(currentUser));

    onAuthStateChanged(auth, (user) => {
      dispatch(signedIn(user));
    });
  };
}

export function forgottenPassword(email) {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_RESET });
    try {
      const auth = getAuth(getState().auth.firebaseApp);
      await sendPasswordResetEmail(auth, email);
      dispatch({ type: RESET_SENT });
      return { type: RESET_SENT };
    } catch (error) {
      dispatch({ type: ERROR_RESET, error });
      return { type: ERROR_RESET };
    }
  };
}
