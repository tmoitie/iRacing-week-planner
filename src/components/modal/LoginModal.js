// @flow
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import { acknowledgeAuthError, createAccount, ERROR_AUTH, signIn, signInWithGoogle } from '../../actions/auth';
import ForgottenPasswordModal from './ForgottenPasswordModal';
import RemoveIcon from '../icon/RemoveIcon';
import Modal from './Modal';
import styles from '../../styles/main.module.scss';
import GoogleIcon from './assets/google.svg';

const errorSelector = (state) => state.auth.errorAuth;
const loadingSelector = (state) => state.auth.loadingAuth;

type Props = {
  isOpen?: boolean,
  onClose: () => void,
};

const defaultProps = {
  isOpen: false,
};

export default function LoginModal({ isOpen, onClose }: Props) {
  const { t } = useTranslation();
  const error = useSelector(errorSelector, shallowEqual);
  const loading = useSelector(loadingSelector);
  const dispatch = useDispatch();
  const [email, setEmail] = React.useState('');
  const [password, setPassword] = React.useState('');
  const [forgotPasswordModalOpen, setFPModalOpen] = React.useState(false);

  const resetForm = () => {
    setEmail('');
    setPassword('');
  };

  const signInClick = async () => {
    const output = await dispatch(signIn(email, password));
    if (output.type !== ERROR_AUTH) {
      resetForm();
    }
  };

  const createAccountClick = async () => {
    const output = await dispatch(createAccount(email, password));
    if (output.type !== ERROR_AUTH) {
      resetForm();
    }
  };

  const googleClick = async () => {
    const output = await dispatch(signInWithGoogle());
    if (output.type !== ERROR_AUTH) {
      resetForm();
    }
  };

  return (
    <Modal
      id="login-modal"
      isOpen={isOpen}
      onClose={onClose}
      title={t('Sign in/Create account')}
      doneAction={onClose}
      showFooter={false}
    >
      <div className={styles['container-fluid']}>
        <p>{t('Signing in will sync your settings across multiple browsers or devices.')}</p>
        <p>
          {t('When you first sign in it will use your current settings to set up the account, but once an account is'
          + ' set-up logging in will overwrite any settings you stored as a guest.')}
        </p>
        {loading ? (
          <div>{t('Loading')}</div>
        ) : (
          <form onSubmit={(e) => {
            e.preventDefault();
            signInClick();
          }}
          >
            {error && (
              <div className={`${styles.alert} ${styles['alert-warning']} ${styles['alert-dismissible']}`} role="alert">
                <button
                  type="button"
                  aria-label="Close"
                  onClick={() => dispatch(acknowledgeAuthError())}
                  className={styles.close}
                >
                  <RemoveIcon />
                </button>
                {t(error.message)}
              </div>
            )}
            <div>
              <button
                type="button"
                className={styles.btn}
                id="googleSignIn"
                onClick={(e) => {
                  e.preventDefault();
                  googleClick();
                }}
              >
                <img src={GoogleIcon} alt="Google logo" style={{ marginInlineEnd: 6 }} />
                {t('Sign in with Google')}
              </button>
            </div>
            <h3>{t('Sign in with email')}</h3>
            <div className={styles['form-group']}>
              <label htmlFor="loginEmail">{t('Email address')}</label>
              <input
                type="email"
                className={styles['form-control']}
                id="loginEmail"
                placeholder={t('Email address')}
                onChange={(e) => setEmail(e.target.value)}
                autoComplete="username"
                value={email}
              />
            </div>
            <div className={styles['form-group']}>
              <label htmlFor="loginPassword">{t('Password')}</label>
              <input
                type="password"
                className={styles['form-control']}
                id="loginPassword"
                placeholder={t('Password')}
                onChange={(e) => setPassword(e.target.value)}
                autoComplete="current-password"
                value={password}
              />
            </div>
            <div className={styles['form-group']}>
              <a
                id="buttonOpenForgottenPassword"
                href=""
                onClick={(e) => { e.preventDefault(); setFPModalOpen(true); }}
              >
                {t('Forgotten password?')}
              </a>
            </div>

            <div>
              <button type="submit" className={`${styles.btn} ${styles['btn-default']}`}>{t('Sign in')}</button>
              <button
                type="button"
                className={styles.btn}
                id="createAccountButton"
                onClick={(e) => {
                  e.preventDefault();
                  createAccountClick();
                }}
                style={{ marginInlineStart: 6 }}
              >
                {t('Create account')}
              </button>
            </div>
          </form>
        )}

      </div>
      <ForgottenPasswordModal
        isOpen={forgotPasswordModalOpen}
        onClose={() => setFPModalOpen(false)}
      />
    </Modal>
  );
}

LoginModal.defaultProps = defaultProps;
