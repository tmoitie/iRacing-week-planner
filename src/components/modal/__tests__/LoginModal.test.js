import { describe, test, expect } from '@jest/globals';
import React from 'react';
import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { getAuth, createUserWithEmailAndPassword, signInWithEmailAndPassword } from 'firebase/auth';

import LoginModal from '../LoginModal';
import { ERROR_ACKNOWLEDGE } from '../../../actions/auth';

jest.mock('firebase/auth');

describe('components/modal/LoginModal', () => {
  const mockStore = configureMockStore([thunk]);

  test('renders closed', async () => {
    const store = mockStore({ auth: { errorAuth: null, loadingAuth: false } });
    render(<Provider store={store}><LoginModal isOpen={false} /></Provider>);
    expect(screen.queryByTestId('login-modal')).not.toBeTruthy();
  });

  test('happy path login', async () => {
    const user = userEvent.setup();
    const firebaseAuth = {};
    getAuth.mockReturnValue(firebaseAuth);
    const store = mockStore({ auth: { errorAuth: null, loadingAuth: false, firebaseApp: {} } });
    render(<Provider store={store}><LoginModal isOpen onClose={() => {}} /></Provider>);

    expect(await screen.findByTestId('login-modal')).toMatchSnapshot();

    await user.type(screen.getByRole('textbox', { name: 'Email address' }), 'tom@example.com');
    await user.type(screen.getByLabelText('Password'), '12345678');

    await user.click(screen.getByText('Sign in'));

    expect(signInWithEmailAndPassword).toHaveBeenCalledWith(firebaseAuth, 'tom@example.com', '12345678');
    expect(await screen.findByTestId('login-modal')).toMatchSnapshot();
  });

  test('login error', async () => {
    const user = userEvent.setup();
    const firebaseAuth = {};
    getAuth.mockReturnValue(firebaseAuth);
    signInWithEmailAndPassword.mockRejectedValueOnce({ message: 'error' });

    const store = mockStore({ auth: { errorAuth: null, loadingAuth: false, firebaseApp: {} } });
    render(<Provider store={store}><LoginModal isOpen onClose={() => {}} /></Provider>);

    await user.click(screen.getByText('Sign in'));
    expect(await screen.findByTestId('login-modal')).toMatchSnapshot();
  });

  test('happy path create account', async () => {
    const user = userEvent.setup();
    const firebaseAuth = {};
    getAuth.mockReturnValue(firebaseAuth);
    const store = mockStore({ auth: { errorAuth: null, loadingAuth: false, firebaseApp: {} } });

    render(<Provider store={store}><LoginModal isOpen onClose={() => {}} /></Provider>);

    expect(await screen.findByTestId('login-modal')).toMatchSnapshot();

    await user.type(screen.getByRole('textbox', { name: 'Email address' }), 'tom@example.com');
    await user.type(screen.getByLabelText('Password'), '12345678');

    await user.click(screen.getByText('Create account'));

    expect(createUserWithEmailAndPassword).toHaveBeenCalledWith(firebaseAuth, 'tom@example.com', '12345678');
    expect(await screen.findByTestId('login-modal')).toMatchSnapshot();
  });

  test('create account error', async () => {
    const user = userEvent.setup();
    const firebaseAuth = {};
    getAuth.mockReturnValue(firebaseAuth);
    createUserWithEmailAndPassword.mockRejectedValueOnce({ message: 'error' });

    const store = mockStore({ auth: { errorAuth: null, loadingAuth: false, firebaseApp: {} } });
    render(<Provider store={store}><LoginModal isOpen onClose={() => {}} /></Provider>);

    await user.click(screen.getByText('Create account'));

    expect(await screen.findByTestId('login-modal')).toMatchSnapshot();
  });

  test('renders error', async () => {
    const user = userEvent.setup();

    const store = mockStore({ auth: { errorAuth: { message: 'Error message' }, loadingAuth: false } });
    const dispatchSpy = jest.spyOn(store, 'dispatch');
    render(<Provider store={store}><LoginModal isOpen onClose={() => {}} /></Provider>);

    expect(await screen.findByTestId('login-modal')).toMatchSnapshot();

    await user.click(within(await screen.findByRole('alert')).getByLabelText('Close'));

    expect(dispatchSpy).toHaveBeenCalledWith({ type: ERROR_ACKNOWLEDGE });
  });

  test('renders loading', async () => {
    const store = mockStore({ auth: { errorAuth: null, loadingAuth: true } });
    render(<Provider store={store}><LoginModal isOpen onClose={() => {}} /></Provider>);

    expect(await screen.findByTestId('login-modal')).toMatchSnapshot();
  });

  test('forgotten password modal control', async () => {
    const user = userEvent.setup();

    const store = mockStore({ auth: { errorAuth: null, loadingAuth: false } });
    render(<Provider store={store}><LoginModal isOpen onClose={() => {}} /></Provider>);

    await user.click(screen.getByText('Forgotten password?'));

    const forgottenModal = await screen.findByTestId('forgottenPasswordModal');
    await user.click(within(forgottenModal).getByLabelText('Close'));

    expect(await screen.findByTestId('login-modal')).toMatchSnapshot();
  });
});
