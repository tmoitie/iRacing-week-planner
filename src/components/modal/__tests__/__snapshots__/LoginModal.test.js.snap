// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/modal/LoginModal create account error 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="login-modal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Sign in/Create account
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Signing in will sync your settings across multiple browsers or devices.
        </p>
        <p>
          When you first sign in it will use your current settings to set up the account, but once an account is set-up logging in will overwrite any settings you stored as a guest.
        </p>
        <form>
          <div>
            <button
              class="btn"
              id="googleSignIn"
              type="button"
            >
              <img
                alt="Google logo"
                src="test-file-stub"
                style="margin-inline-end: 6px;"
              />
              Sign in with Google
            </button>
          </div>
          <h3>
            Sign in with email
          </h3>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              autocomplete="username"
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="loginPassword"
            >
              Password
            </label>
            <input
              autocomplete="current-password"
              class="form-control"
              id="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <a
              href=""
              id="buttonOpenForgottenPassword"
            >
              Forgotten password?
            </a>
          </div>
          <div>
            <button
              class="btn btn-default"
              type="submit"
            >
              Sign in
            </button>
            <button
              class="btn"
              id="createAccountButton"
              style="margin-inline-start: 6px;"
              type="button"
            >
              Create account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/LoginModal forgotten password modal control 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="login-modal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Sign in/Create account
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Signing in will sync your settings across multiple browsers or devices.
        </p>
        <p>
          When you first sign in it will use your current settings to set up the account, but once an account is set-up logging in will overwrite any settings you stored as a guest.
        </p>
        <form>
          <div>
            <button
              class="btn"
              id="googleSignIn"
              type="button"
            >
              <img
                alt="Google logo"
                src="test-file-stub"
                style="margin-inline-end: 6px;"
              />
              Sign in with Google
            </button>
          </div>
          <h3>
            Sign in with email
          </h3>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              autocomplete="username"
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="loginPassword"
            >
              Password
            </label>
            <input
              autocomplete="current-password"
              class="form-control"
              id="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <a
              href=""
              id="buttonOpenForgottenPassword"
            >
              Forgotten password?
            </a>
          </div>
          <div>
            <button
              class="btn btn-default"
              type="submit"
            >
              Sign in
            </button>
            <button
              class="btn"
              id="createAccountButton"
              style="margin-inline-start: 6px;"
              type="button"
            >
              Create account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/LoginModal happy path create account 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="login-modal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Sign in/Create account
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Signing in will sync your settings across multiple browsers or devices.
        </p>
        <p>
          When you first sign in it will use your current settings to set up the account, but once an account is set-up logging in will overwrite any settings you stored as a guest.
        </p>
        <form>
          <div>
            <button
              class="btn"
              id="googleSignIn"
              type="button"
            >
              <img
                alt="Google logo"
                src="test-file-stub"
                style="margin-inline-end: 6px;"
              />
              Sign in with Google
            </button>
          </div>
          <h3>
            Sign in with email
          </h3>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              autocomplete="username"
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="loginPassword"
            >
              Password
            </label>
            <input
              autocomplete="current-password"
              class="form-control"
              id="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <a
              href=""
              id="buttonOpenForgottenPassword"
            >
              Forgotten password?
            </a>
          </div>
          <div>
            <button
              class="btn btn-default"
              type="submit"
            >
              Sign in
            </button>
            <button
              class="btn"
              id="createAccountButton"
              style="margin-inline-start: 6px;"
              type="button"
            >
              Create account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/LoginModal happy path create account 2`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="login-modal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Sign in/Create account
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Signing in will sync your settings across multiple browsers or devices.
        </p>
        <p>
          When you first sign in it will use your current settings to set up the account, but once an account is set-up logging in will overwrite any settings you stored as a guest.
        </p>
        <form>
          <div>
            <button
              class="btn"
              id="googleSignIn"
              type="button"
            >
              <img
                alt="Google logo"
                src="test-file-stub"
                style="margin-inline-end: 6px;"
              />
              Sign in with Google
            </button>
          </div>
          <h3>
            Sign in with email
          </h3>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              autocomplete="username"
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="loginPassword"
            >
              Password
            </label>
            <input
              autocomplete="current-password"
              class="form-control"
              id="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <a
              href=""
              id="buttonOpenForgottenPassword"
            >
              Forgotten password?
            </a>
          </div>
          <div>
            <button
              class="btn btn-default"
              type="submit"
            >
              Sign in
            </button>
            <button
              class="btn"
              id="createAccountButton"
              style="margin-inline-start: 6px;"
              type="button"
            >
              Create account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/LoginModal happy path login 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="login-modal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Sign in/Create account
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Signing in will sync your settings across multiple browsers or devices.
        </p>
        <p>
          When you first sign in it will use your current settings to set up the account, but once an account is set-up logging in will overwrite any settings you stored as a guest.
        </p>
        <form>
          <div>
            <button
              class="btn"
              id="googleSignIn"
              type="button"
            >
              <img
                alt="Google logo"
                src="test-file-stub"
                style="margin-inline-end: 6px;"
              />
              Sign in with Google
            </button>
          </div>
          <h3>
            Sign in with email
          </h3>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              autocomplete="username"
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="loginPassword"
            >
              Password
            </label>
            <input
              autocomplete="current-password"
              class="form-control"
              id="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <a
              href=""
              id="buttonOpenForgottenPassword"
            >
              Forgotten password?
            </a>
          </div>
          <div>
            <button
              class="btn btn-default"
              type="submit"
            >
              Sign in
            </button>
            <button
              class="btn"
              id="createAccountButton"
              style="margin-inline-start: 6px;"
              type="button"
            >
              Create account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/LoginModal happy path login 2`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="login-modal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Sign in/Create account
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Signing in will sync your settings across multiple browsers or devices.
        </p>
        <p>
          When you first sign in it will use your current settings to set up the account, but once an account is set-up logging in will overwrite any settings you stored as a guest.
        </p>
        <form>
          <div>
            <button
              class="btn"
              id="googleSignIn"
              type="button"
            >
              <img
                alt="Google logo"
                src="test-file-stub"
                style="margin-inline-end: 6px;"
              />
              Sign in with Google
            </button>
          </div>
          <h3>
            Sign in with email
          </h3>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              autocomplete="username"
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="loginPassword"
            >
              Password
            </label>
            <input
              autocomplete="current-password"
              class="form-control"
              id="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <a
              href=""
              id="buttonOpenForgottenPassword"
            >
              Forgotten password?
            </a>
          </div>
          <div>
            <button
              class="btn btn-default"
              type="submit"
            >
              Sign in
            </button>
            <button
              class="btn"
              id="createAccountButton"
              style="margin-inline-start: 6px;"
              type="button"
            >
              Create account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/LoginModal login error 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="login-modal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Sign in/Create account
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Signing in will sync your settings across multiple browsers or devices.
        </p>
        <p>
          When you first sign in it will use your current settings to set up the account, but once an account is set-up logging in will overwrite any settings you stored as a guest.
        </p>
        <form>
          <div>
            <button
              class="btn"
              id="googleSignIn"
              type="button"
            >
              <img
                alt="Google logo"
                src="test-file-stub"
                style="margin-inline-end: 6px;"
              />
              Sign in with Google
            </button>
          </div>
          <h3>
            Sign in with email
          </h3>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              autocomplete="username"
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="loginPassword"
            >
              Password
            </label>
            <input
              autocomplete="current-password"
              class="form-control"
              id="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <a
              href=""
              id="buttonOpenForgottenPassword"
            >
              Forgotten password?
            </a>
          </div>
          <div>
            <button
              class="btn btn-default"
              type="submit"
            >
              Sign in
            </button>
            <button
              class="btn"
              id="createAccountButton"
              style="margin-inline-start: 6px;"
              type="button"
            >
              Create account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/LoginModal renders error 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="login-modal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Sign in/Create account
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Signing in will sync your settings across multiple browsers or devices.
        </p>
        <p>
          When you first sign in it will use your current settings to set up the account, but once an account is set-up logging in will overwrite any settings you stored as a guest.
        </p>
        <form>
          <div
            class="alert alert-warning alert-dismissible"
            role="alert"
          >
            <button
              aria-label="Close"
              class="close"
              type="button"
            >
              <span
                class="glyphicon glyphicon-remove"
              />
            </button>
            Error message
          </div>
          <div>
            <button
              class="btn"
              id="googleSignIn"
              type="button"
            >
              <img
                alt="Google logo"
                src="test-file-stub"
                style="margin-inline-end: 6px;"
              />
              Sign in with Google
            </button>
          </div>
          <h3>
            Sign in with email
          </h3>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              autocomplete="username"
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="loginPassword"
            >
              Password
            </label>
            <input
              autocomplete="current-password"
              class="form-control"
              id="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <a
              href=""
              id="buttonOpenForgottenPassword"
            >
              Forgotten password?
            </a>
          </div>
          <div>
            <button
              class="btn btn-default"
              type="submit"
            >
              Sign in
            </button>
            <button
              class="btn"
              id="createAccountButton"
              style="margin-inline-start: 6px;"
              type="button"
            >
              Create account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/LoginModal renders loading 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="login-modal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Sign in/Create account
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Signing in will sync your settings across multiple browsers or devices.
        </p>
        <p>
          When you first sign in it will use your current settings to set up the account, but once an account is set-up logging in will overwrite any settings you stored as a guest.
        </p>
        <div>
          Loading
        </div>
      </div>
    </div>
  </div>
</div>
`;
